src/act.ts

```diff
@@ -1,7 +1,101 @@
-import { act } from 'react-test-renderer';
+import { act as reactTestRendererAct } from 'react-test-renderer';
+import { checkReactVersionAtLeast } from './checkReactVersionAtLeast';
 
 const actMock = (callback: () => void) => {
   callback();
 };
 
-export default act || actMock;
+type GlobalWithReactActEnvironment = {
+  IS_REACT_ACT_ENVIRONMENT?: boolean;
+} & typeof globalThis;
+function getGlobalThis(): GlobalWithReactActEnvironment {
+  // eslint-disable-next-line no-restricted-globals
+  if (typeof self !== 'undefined') {
+    // eslint-disable-next-line no-restricted-globals
+    return self as GlobalWithReactActEnvironment;
+  }
+  if (typeof window !== 'undefined') {
+    return window;
+  }
+  if (typeof global !== 'undefined') {
+    return global;
+  }
+
+  throw new Error('unable to locate global object');
+}
+
+function setIsReactActEnvironment(isReactActEnvironment: boolean | undefined) {
+  getGlobalThis().IS_REACT_ACT_ENVIRONMENT = isReactActEnvironment;
+}
+
+function getIsReactActEnvironment() {
+  return getGlobalThis().IS_REACT_ACT_ENVIRONMENT;
+}
+
+type Act = typeof reactTestRendererAct;
+function withGlobalActEnvironment(actImplementation: Act) {
+  return (callback: Parameters<Act>[0]) => {
+    const previousActEnvironment = getIsReactActEnvironment();
+    setIsReactActEnvironment(true);
+
+    // this code is riddled with eslint disabling comments because this doesn't use real promises but eslint thinks we do
+    try {
+      // The return value of `act` is always a thenable.
+      let callbackNeedsToBeAwaited = false;
+      const actResult = actImplementation(() => {
+        const result = callback();
+        if (
+          result !== null &&
+          typeof result === 'object' &&
+          // eslint-disable-next-line promise/prefer-await-to-then
+          typeof (result as any).then === 'function'
```

Comment:
Maybe that will work?
Or 
```
// @ts-expect-error this should be a promise or thenable
```
```suggestion
          typeof result['then'] === 'function'
```

Reply:
First one doesn't work, but expecting the error is indeed good